File: algorithms/java/src/test/java/leetcode/LC331VerifyPreorderSerializationOfABinaryTree.java
package leetcode.lc331_verify_preorder_serialization_of_a_binary_tree;

import org.junit.jupiter.api.Test;


import java.util.Stack;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/verify-preorder-serialization-of-a-binary-tree/
 * #medium
 */
public final class LC331VerifyPreorderSerializationOfABinaryTree {
    public final class Solution {
        public boolean isValidSerialization(final String preorder) {
            if (null == preorder || preorder.length() == 0) return false;
            String[] items = preorder.split(",");
            Stack<Integer> kids = new Stack<>();
            kids.push(1);
            for (String node : items) {
                while (kids.peek() == 2) {
                    kids.pop();
                    if (kids.size() == 0) return false;
                }
                kids.push(kids.pop() + 1);
                if (!node.equals("#")) {
                    kids.push(0);
                }
            }
            while (kids.size() != 0 && kids.peek() == 2) kids.pop();
            return kids.size() == 0;
        }
    }

    @Test
    public void test_empty() throws Exception {
        assertTrue(new Solution().isValidSerialization("#"));
    }

    @Test
    public void test_empty_but_not_really() throws Exception {
        assertFalse(new Solution().isValidSerialization("#1"));
    }

    @Test
    public void test_example_1() throws Exception {
        assertTrue(new Solution().isValidSerialization("9,3,4,#,#,1,#,#,2,#,6,#,#"));
    }

    @Test
    public void test_example_2() throws Exception {
        assertFalse(new Solution().isValidSerialization("1,#"));
    }

    @Test
    public void test_example_3() throws Exception {
        assertFalse(new Solution().isValidSerialization("9,#,#,1"));
    }
}

File: algorithms/code/leetcode/lc331_verify_preorder_serialization_of_a_binary_tree/lc331_verify_preorder_serialization_of_a_binary_tree.py
#!/usr/bin/env python3
# https://leetcode.com/problems/verify-preorder-serialization-of-a-binary-tree/

import unittest


class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        if None is preorder or len(preorder) == 0:
            return False
        items = preorder.split(',')
        kids = [1]
        for node in items:
            while kids[-1] == 2:
                kids.pop()
                if len(kids) == 0:
                    return False
            kids.append(kids.pop() + 1)
            if node != '#':
                kids.append(0)
        while len(kids) != 0 and kids[-1] == 2:
            kids.pop()
        return len(kids) == 0


class TestCode(unittest.TestCase):
    def test_empty(self):
        self.assertTrue(Solution().isValidSerialization('#'))

    def test_empty_but_not_really(self):
        self.assertFalse(Solution().isValidSerialization('#1'))

    def test_example_1(self):
        self.assertTrue(
            Solution().isValidSerialization('9,3,4,#,#,1,#,#,2,#,6,#,#')
        )

    def test_example_2(self):
        self.assertFalse(Solution().isValidSerialization('1,#'))

    def test_example_3(self):
        self.assertFalse(Solution().isValidSerialization('9,#,#,1'))

    def test_nothing(self):
        self.assertFalse(Solution().isValidSerialization(''))
